handshake/src/types/lockingscript.rs

```diff
@@ -0,0 +1,263 @@
+//! Handshake LockingScript and WitnessProgram
+
+use std::io::{Read, Write};
+
+use coins_core::{
+    ser::ByteFormat,
+    types::tx::RecipientIdentifier
+};
+
+wrap_prefixed_byte_vector!(
+    /// A WitnessProgram represents the data field of a LockingScript.
+    /// Since Handshake is segwit only, the WitnessProgram doesn't contain
+    /// opcodes itself, it is templated into a script at runtime. The
+    /// size of the WitnessProgram determines how it is interpreted for
+    /// version 0 LockingScripts.
+    WitnessProgram
+);
+
+impl From<[u8; 20]> for WitnessProgram {
+    fn from(v: [u8; 20]) -> Self {
+        let mut hash = vec![0x14];
+        hash.append(&mut v.to_vec());
+        WitnessProgram::new(hash)
+    }
+}
+
+impl From<WitnessProgram> for [u8; 20] {
+    fn from(w: WitnessProgram) -> [u8; 20] {
+        let mut data = [0; 20];
+        data.clone_from_slice(&w.0[..]);
+        data
+    }
+}
+
+impl From<WitnessProgram> for [u8; 32] {
+    fn from(w: WitnessProgram) -> [u8; 32] {
+        let mut data = [0; 32];
+        data.clone_from_slice(&w.0[..]);
+        data
+    }
+}
+
+// TODO: test this
+impl From<WitnessProgram> for Vec<u8> {
+    fn from(w: WitnessProgram) -> Vec<u8> {
+        w.0[1..].to_vec()
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq)]
+/// A LockingScript
+pub struct LockingScript {
+    /// The version field determines how the Witness Program is interpreted
+    /// by the virtual machine.
+    pub version: u8,
+    /// The witness_program is generally a committment to some data that is required
+    /// for virtual machine execution. For version 0, a witness_program that is 20
+    /// bytes is interpreted as a pay-to-witness-pubkeyhash and data that is 32 bytes
+    /// is interpreted as a pay-to-witness-scripthash.
+    pub witness_program: WitnessProgram
+}
+
+impl LockingScript {
+    /// Returns a null LockingScript
+    pub fn null() -> Self {
+        Self {
+            version: 0,
+            witness_program: WitnessProgram(vec![0x00, 20])
+        }
+    }
+
+    /// Create a new LockingScript
+    pub fn new(v: Vec<u8>) -> Self {
+        let (version_and_size, data) = v.split_at(2);
+        assert_eq!(version_and_size[1], data.len() as u8);
```

Comment:
I think that this `assert_eq!` should be removed and instead have the function return a `Result<Self, Error>`